readalongs/web_api.py

```diff
@@ -203,5 +204,174 @@ def create_grammar(xml):
+
+    Encoding: all input and output is in UTF-8.
+
+    Args (as dict items in the request body):
+     - audio_duration: duration in seconds of the audio file used to create the alignment
+     - output_format: one of TextGrid, eaf, srt, vtt
+     - xml: the XML file produced by /assemble
+     - smil: the SMIL file produced by SoundSwallower(.js)
+
+    Formats supported:
+     - TextGrid: Praat TextGrid file format
+     - eaf: ELAN eaf file format
+     - srt: SRT subtitle format, as two files, 1) for sentences, 2) for words
+     - vtt: WebVTT subtitle format, as two files, 1) for sentences, 2) for words
+
+    Data privacy consideration: due to limitations of the libraries used to perform
+    some of these conversions, the output files will be temporarily stored on disk,
+    but they get deleted immediately as this endpoint returns its output or reports
+    any error.
+
+    Returns:
+     - file_name: a suggested name for the returned file
+     - file_contents: the contents of the file converted in the requested format
+     for srt and vtt:
+     - other_file_name: a suggested name for the second file
+     - other_file_contents: the contents of the second file
+    """
+    try:
+        parsed_xml = etree.fromstring(bytes(input.xml, encoding="utf-8"))
+    except etree.XMLSyntaxError as e:
+        raise HTTPException(status_code=422, detail="XML provided is not valid") from e
+
+    try:
+        words = parse_smil(input.smil)
+    except ValueError as e:
+        raise HTTPException(status_code=422, detail="SMIL provided is not valid") from e
+
+    # Data privacy consideration: creating the temporary directory with a context
+    # manager like this guarantees, as we promise in the API documentation, that the
+    # temporary directory and all temporary files we create in it will be deleted
+    # when this function exits, whether it is with an error or with success.
+    with TemporaryDirectory() as temp_dir_name:
+        prefix = os.path.join(temp_dir_name, "f")
+
+        output_format = input.output_format.lower()
+        if output_format == "textgrid":
+            save_label_files(
+                words, parsed_xml, input.audio_duration, prefix, "textgrid"
+            )
+            return ConvertResponse(
+                file_name="aligned.TextGrid",
+                file_contents=slurp_file(prefix + ".TextGrid"),
+            )
+
+        elif output_format == "eaf":
+            save_label_files(words, parsed_xml, input.audio_duration, prefix, "eaf")
+            return ConvertResponse(
+                file_name="aligned.eaf",
+                file_contents=slurp_file(prefix + ".eaf"),
+            )
+
+        elif output_format == "srt":
+            save_subtitles(words, parsed_xml, prefix, "srt")
+            return ConvertResponse(
+                file_name="aligned_sentences.srt",
+                file_contents=slurp_file(prefix + "_sentences.srt"),
+                other_file_name="aligned_words.srt",
```

Comment:
So yes, just to re-iterate, my recommendations here would be: 

1. return `FileResponse` objects instead of this custom `ConvertResponse` object
2. use path parameters for each output format and use a Python Enumeration to make use of FastAPI handling exceptions
3. use query parameters to return either sentence or word subtitles for subtitle formats (with sentences being the default as this is what the vast majority of cases will want I expect)